CAIPs/caip-stream-sync.md

```diff
@@ -0,0 +1,193 @@
+
+### Protocol messages
+
+The stream sync protocol has four main message types that nodes send to each other. It’s implemented as a libp2p sub-protocol, and is inspired by [ipfs bitswap](https://docs.ipfs.tech/concepts/bitswap/).
+
+**Libp2p protocol id:**
+
+```
+/caip/streamsync/1.0.0
+```
+
+#### Synchronizing tips
+
+The following two protocol messages are used to query and gossip tips.
+
+**SubscribeSet**
+
+This message is used to subscribe to a set of streams. 
+
+```ipldsch
+type SubscribeSet struct {
+  separatorKey nullable String
+  separatorValue nullable Bytes
+  streamFilter nullable Bytes
+  controllerFilter nullable Bytes
+} representation tuple
+```
+
+*Properties:*
+
+* `separatorKey` - a string key that has to be set in the *genesis event header* of a matching stream
+* `separatorValue` - the value set for the `separatorKey`
+* `streamFilter` - a byte array which streamids in this set must end with, e.g. if this property contains 3 bytes all streamids that ends with these three bytes are part of the set
+* `controllerFilter` - A bloom filter populated with the DIDs that the set includes
+
+*Invariants:*
+
+* If `separatorKey` and `separatorValue` are null, `streamFilter` must be a full streamId
+* If `separatorKey` and `separatorValue` are set, `shardFilter` and `ctrlFilter` are optional
+
+**HaveTips**
+
+A response to a *SubscribeSet* message or an update sent to known subscribed peers containing the latest known tips of the given stream set.
+
+```ipldsch
+type HaveTips struct {
+  tips [Tips]
+  separatorKey nullable String
+  separatorValue nullable Bytes
+  streamFilter nullable Bytes
+  ctrlFilter nullable Bytes
+} representation tuple
+
+type Tips struct {
+  streamid Bytes
+  tips [CID]
+} representation tuple
+```
+
+*Properties:*
+
+* `tips` - all the tips that the node has for this set
+  * `streamid` - the id of the corresponding tips
+  * `tips` - a list of latest tips
+
+* `separatorKey`, `separatorValue`, `streamFilter`, and `controllerFilter` - same as in *SubscribeSet*
+
+#### Synchronizing Events
+
+The following two protocol messages are used to synchronize event data between peers.
+
+**WantStream**
+
+Request to stream events from another node given a streamId.
+
+```ipldsch
+type WantStream struct {
+  streamid Bytes
+  knownTips [CID] // the CID(s) of the event to start the streaming from
+  offset nullable Integer // number of bytes already received
+} representation tuple
+```
+
+*Properties:*
+
+* `streamid` - the streamId being requested
+* `knownTips` - the CIDs of the tips that the requester already has locally
+* `offset` - number of bytes already recieved on top of the known tip
```

Comment:
"on top of the known tip", or since the beginning of the stream from genesis event?